### ISSUE TITLE
Failing ROS2 comms on QEMU

### ISSUE THREAD

**Issue by dodsonmg** [eclipse-cyclonedds/cyclonedds#488]
TL;DR//I am trying to troubleshoot what appear to be communication issues running ROS2 on QEMU.  I tried to build and run the cyclonedds examples, to eliminate at least one layer as my problem source, but the Maven dependency kept me from doing so.

---

Background:

I am trying to cross-compile ROS2 demos on a FreeBSD-based research system running on QEMU.  I am using the cyclonedds middleware.

My QEMU is configured to allow communication from guest -> host, but *not* host -> guest.  If I run a simple publisher on the guest, the host can see the node, but can't query it, which makes sense.  What makes less sense (though this may be a conceptual error on my part), the host can't see the topics to which the publisher is publishing.

If I attempt to run a publisher and subscriber separately, but simultaneously, on the guest (to avoid trying to discover and communicate between guest and host), I get the following error upon starting the second process:

```
1586640423.685765 [42]     100051: UDP make_socket failed for multicast port 17900
[ERROR] [1586640423.781729296] [rmw_cyclonedds_cpp]: rmw_create_node: failed to create domain, error Error

>>> [rcutils|error_handling.c:108] rcutils_set_error_state()
This error state is being overwritten:

  'error not set, at /home/broomstick/cheri/ros2_min/src/ros2/rcl/rcl/src/rcl/node.c:336'

with this new error message:

  'rcl node's rmw handle is invalid, at /home/broomstick/cheri/ros2_min/src/ros2/rcl/rcl/src/rcl/node.c:485'

rcutils_reset_error() should be called after error handling to avoid this.
<<<
[ERROR] [1586640423.895518056] [rcl]: Failed to fini publisher for node: 1
Terminating due to uncaught exception 0x48960300 of type rclcpp::exceptions::RCLError
Abort
```

When I try to run a composed publisher and subscriber (two ROS2 nodes in the same process), I get a slightly different error.  As you can see, the subscriber, which should simply echo the publisher's message, echos only an empty string.

```
[INFO] [1586660033.578303216] [publisher_node]: Publisher: 'Hello, world! 0'

>>> [rcutils|error_handling.c:108] rcutils_set_error_state()
This error state is being overwritten:

  'Security was requested but the Cyclone DDS being used does not have security support enabled. Recompile Cyclone DDS with the '-DENABLE_SECURITY=ON' CMake option, at /home/broomstick/cheri/ros2_min/src/ros2/rmw_cyclonedds/rmw_cyclonedds_cpp/src/rmw_node.cpp:834'

with this new error message:

  'invalid data size, at /home/broomstick/cheri/ros2_min_no_cheri/src/ros2/rmw_cyclonedds/rmw_cyclonedds_cpp/src/serdata.cpp:308'

rcutils_reset_error() should be called after error handling to avoid this.
<<<

[INFO] [1586660033.973517976] [subscriber_node]: Subscriber: ''
```
---

Problem: 

I wanted to eliminate the ROS middleware interfaces from my troubleshooting by building the cyclonedds examples, but these appear to have a dependency on Maven, based on the following snippets, which I can't currently cross-compile for my target system.

https://github.com/eclipse-cyclonedds/cyclonedds/blob/b46bd7ffff53a86375f36010049bbc19294e7d0d/CMakeLists.txt#L217-L219

and 

https://github.com/eclipse-cyclonedds/cyclonedds/blob/b46bd7ffff53a86375f36010049bbc19294e7d0d/CMakeLists.txt#L51-L55

---

Additional information:

The research system we're using limits operations on pointers, so I've removed some of the atomic operations, such as the one in the snippet below.  None of these functions appear to be called anywhere in my code base.

https://github.com/eclipse-cyclonedds/cyclonedds/blob/b46bd7ffff53a86375f36010049bbc19294e7d0d/src/ddsrt/include/dds/ddsrt/atomics/gcc.h#L85-L87

---

I would be grateful for any insight into either of the following:

1. The errors I'm receiving
2. Simplified building of the cyclonedds examples

**Comment by eboasson** [comment]
> TL;DR

Actually I'm intrigued and read it in full 🙂 You wouldn't happen to have some links to papers with more details about the research system?

> I am trying to cross-compile ROS2 demos on a FreeBSD-based research system running on QEMU. I am using the cyclonedds middleware.

I've never tried to run it on FreeBSD, but I'm quite sure any obstacles will prove to be pretty trivial (in hindsight anyway).

> My QEMU is configured to allow communication from guest -> host, but _not_ host -> guest. If I run a simple publisher on the guest, the host can see the node, but can't query it, which makes sense. What makes less sense (though this may be a conceptual error on my part), the host can't see the topics to which the publisher is publishing.

Everything really is based on two-way communication at the protocol level. Nodes need to discover each other, send data & acknowledgements, &c. This is very similar to the needs of TCP, the key difference is that TCP isn't peer-to-peer but client-to-server. That means you can allow "outbound" connections only, which really means that you can initiate connections in only one direction. Within the established connection, you do get two-way traffic.

I wouldn't expect DDS (or, more precisely, DDSI-RTPS over UDP) to work without bidirectional communication.
> If I attempt to run a publisher and subscriber separately, but simultaneously, on the guest (to avoid trying to discover and communicate between guest and host), I get the following error upon starting the second process:
> 
> ```
> 1586640423.685765 [42]     100051: UDP make_socket failed for multicast port 17900

I very much suspect that to be https://github.com/eclipse-cyclonedds/cyclonedds/blob/b46bd7ffff53a86375f36010049bbc19294e7d0d/src/ddsrt/src/sockets/posix/socket.c#L319-L325, that it should also do this for FreeBSD.

> When I try to run a composed publisher and subscriber (two ROS2 nodes in the same process), I get a slightly different error. As you can see, the subscriber, which should simply echo the publisher's message, echos only an empty string.
> 
> ```
> [INFO] [1586660033.578303216] [publisher_node]: Publisher: 'Hello, world! 0'
> 
> >>> [rcutils|error_handling.c:108] rcutils_set_error_state()
> This error state is being overwritten:
> 
>   'Security was requested but the Cyclone DDS being used does not have security support enabled. Recompile Cyclone DDS with the '-DENABLE_SECURITY=ON' CMake option, at /home/broomstick/cheri/ros2_min/src/ros2/rmw_cyclonedds/rmw_cyclonedds_cpp/src/rmw_node.cpp:834'
> 
> with this new error message:
> 
>   'invalid data size, at /home/broomstick/cheri/ros2_min_no_cheri/src/ros2/rmw_cyclonedds/rmw_cyclonedds_cpp/src/serdata.cpp:308'

That line is where it ends up if deserialising the message failed. It'd be interesting to see whether it is CDR encoded data that's the problem or the deserialization. Quite a few message types get tested and so this might be something specific to the platform.

> Problem:
> 
> I wanted to eliminate the ROS middleware interfaces from my troubleshooting by building the cyclonedds examples, but these appear to have a dependency on Maven, based on the following snippets, which I can't currently cross-compile for my target system.

I apologize for the Maven and Java dependency ... there is a way out: the only thing Java and Maven are used for is generating the IDL compiler, but the output of it is completely system independent. So if you're not afraid of a hack, take you could take the build/src/tools/ddsperf/ddsperf_types.c and .h generated on some other platform, copy them into src/tools/ddsperf and change the CMakeFiles to build ddsperf regardless of BUILD_IDC and to use the files generated elsewhere rather then doing the "idlc_generate" thing.

The same holds for the examples, but ddsperf can very easily be used as a sanity-checking device ("ddsperf -h" gives some information, but here I'm especially thinking of invoking it as "ddsperf sanity" on both sides).

I have on occasion considered doing the pre-generate thing for the IDL-to-C for at least "ddsperf". If you run into difficulties let me know.

> Additional information:
> 
> The research system we're using limits operations on pointers, so I've removed some of the atomic operations, such as the one in the snippet below. None of these functions appear to be called anywhere in my code base.
> 
> https://github.com/eclipse-cyclonedds/cyclonedds/blob/b46bd7ffff53a86375f36010049bbc19294e7d0d/src/ddsrt/include/dds/ddsrt/atomics/gcc.h#L85-L87

That's the category "it's fine as long as it builds" 🙂

**Comment by dodsonmg** [comment]
I modified the code snippet in `socket.c` to be `#if defined(__APPLE__) || defined(__FreeBSD__)` and it does, indeed, eliminate the `UDP make_socket failed` error.

I also added FreeBSD to the guard here, though it's less clear to me that it's necessary:
https://github.com/eclipse-cyclonedds/cyclonedds/blob/b46bd7ffff53a86375f36010049bbc19294e7d0d/src/ddsrt/src/sockets/posix/socket.c#L35-L37

While I no longer receive the error, when I run both a publisher and subscriber on the QEMU guest together, the subscriber still doesn't pick anything up.  I get output such as the following:

```
[INFO] [publisher_node]: Publisher: 'Hello, world! 0'                                                                                                                                                       
[INFO] [subscriber_node]: Subscriber: ''                                                                                                                                                                    
[INFO] [publisher_node]: Publisher: 'Hello, world! 1'                                                                                                                                                       
[INFO] [subscriber_node]: Subscriber: ''         
```

Similarly, as before, the host sees both the publisher and subscriber nodes, but not the topic.  E.g., `ros2 node list` shows both nodes, but `ros2 topic list` does not show the topic to which the publisher is trying to publish.

I can submit a PR to add FreeBSD.

---

I haven't gone back to the examples yet, but I'll hack around, as you suggested, and report back here.

---

The target research system is called CheriBSD, and it uses hardware-enforced capabilities (similar to fat pointers) to efficiently enforce pointer provenance, pointer bounding, and control flow.  The overall project is called CHERI, and you can find a good, summary description as well as links to papers and tech reports [here](https://www.cl.cam.ac.uk/research/security/ctsrd/cheri/).

This look at the Robot Operating System is our first poke at the cyber-physical domain, and I don't have any documentation to show you, but hopefully that will come in time.  CHERI provides some really intriguing prospects for simplified and efficient compartmentalisation and access control, in addition to it's memory safety guarantees, which is my area of interest.

**Comment by k0ekk0ek** [comment]
I won't interfere too much, but wanted to state that a PR to add FreeBSD would be very much appreciated :+1:. Since Travis isn't exactly what it used to be I've been contemplating moving the builds to Cirrus (I use it for another project) and they support FreeBSD on their CI platform too (as well as up-to-date Windows hosts).

**Comment by dodsonmg** [comment]
> I apologize for the Maven and Java dependency ... there is a way out: the only thing Java and Maven are used for is generating the IDL compiler, but the output of it is completely system independent. So if you're not afraid of a hack, take you could take the build/src/tools/ddsperf/ddsperf_types.c and .h generated on some other platform, copy them into src/tools/ddsperf and change the CMakeFiles to build ddsperf regardless of BUILD_IDC and to use the files generated elsewhere rather then doing the "idlc_generate" thing.

I built ddsperf on the host (running Ubuntu 18.04) and tested `ddsperf sanity` in two terminals, which appeared to give me sane results.

I then copied `ddsperf_type.c` and `ddsperf_type.h` from `<host_prefix>/build/cyclonedds/src/tools/ddsperf` to `<target_prefix>/src/cyclonedds/src/tools/ddsperf` and modified the `CMakeList.txt` files in `tools` and `ddsperf`.

The `tools` modification removed the conditional around building `ddsperf`:
https://github.com/eclipse-cyclonedds/cyclonedds/blob/b46bd7ffff53a86375f36010049bbc19294e7d0d/src/tools/CMakeLists.txt#L15-L17

The `ddsperf` modification removed the `idlc_generate` call:
https://github.com/eclipse-cyclonedds/cyclonedds/blob/b46bd7ffff53a86375f36010049bbc19294e7d0d/src/tools/ddsperf/CMakeLists.txt#L13-L15

I then went ahead rebuilding cyclonedds with colcon and got the following error:
```
ld.lld: error: unable to find library -lddsperf_types
clang-11: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [bin/ddsperf] Error 1
make[1]: *** [src/tools/ddsperf/CMakeFiles/ddsperf.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
make: *** [all] Error 2
---
Failed   <<< cyclonedds [ Exited with code 2 ]
```

I assume this is related to the `target_link_libraries(ddsperf ddsperf_types ddsc)` in `ddsperf/CMakeList.txt:15` above, but I'm not sufficiently competent to figure out how to get around it.

On the host, building ddsperf also created the following:
```
<host_prefix>/build/cyclonedds/src/tools/ddsperf/CMakeFiles/ddsperf_types_idlc_generate.dir
<host_prefix>/build/cyclonedds/src/tools/ddsperf/CMakeFiles/ddsperf.dir/ddsperf_types.c.o
```

I'm not sure if that's relevant, but since I'm trying to do a clean build of cyclonedds, they don't exist on the target before I attempt the build.  I guess if they're necessary, I can try some kind of bootstrapping...

**Comment by eboasson** [comment]
> I then went ahead rebuilding cyclonedds with colcon and got the following error:
> 
> ```
> ld.lld: error: unable to find library -lddsperf_types
> clang-11: error: linker command failed with exit code 1 (use -v to see invocation)
> make[2]: *** [bin/ddsperf] Error 1
> make[1]: *** [src/tools/ddsperf/CMakeFiles/ddsperf.dir/all] Error 2
> make[1]: *** Waiting for unfinished jobs....
> make: *** [all] Error 2
> ---
> Failed   <<< cyclonedds [ Exited with code 2 ]
> ```
> 
> I assume this is related to the `target_link_libraries(ddsperf ddsperf_types ddsc)` in `ddsperf/CMakeList.txt:15` above, but I'm not sufficiently competent to figure out how to get around it.

That's indeed the reason it fails. I pushed a commit on a separate branch in my private fork that should do the trick: https://github.com/eboasson/cyclonedds/commit/a63a433727f659aaf00b0d8bd8db9f988e175da0

it is the same as what you did, but with one additional step: I added the ddsperf_types.c and .h files to the sources of ddsperf and removed the reference to the library.

**Comment by eboasson** [comment]
> While I no longer receive the error, when I run both a publisher and subscriber on the QEMU guest together, the subscriber still doesn't pick anything up. I get output such as the following:
> 
> ```
> [INFO] [publisher_node]: Publisher: 'Hello, world! 0'                                                                                                                                                       
> [INFO] [subscriber_node]: Subscriber: ''                                                                                                                                                                    
> [INFO] [publisher_node]: Publisher: 'Hello, world! 1'                                                                                                                                                       
> [INFO] [subscriber_node]: Subscriber: ''         
> ```
> 
> Similarly, as before, the host sees both the publisher and subscriber nodes, but not the topic. E.g., `ros2 node list` shows both nodes, but `ros2 topic list` does not show the topic to which the publisher is trying to publish.

Does the same test work fine on Ubuntu? I imagine so. This behaviour matches a failure to deserialise the message, which also what the "'invalid data size, at /home/broomstick/cheri/ros2_min_no_cheri/src/ros2/rmw_cyclonedds/rmw_cyclonedds_cpp/src/serdata.cpp:308'" exception you reported earlier indicates.

I could be a FreeBSD vs the others little detail with memory layouts, or a mishap in the endianness used for encoding the string size. If you can do a packet capture with Wireshark (or you could let Cyclone write a capture file by setting the "Tracing/PacketCaptureFile" option, you might want to refer to https://github.com/eclipse-cyclonedds/cyclonedds#configuration) we could check whether the problem is in the serialisation of the string, or rather in the interpretation.

**Comment by dodsonmg** [comment]
> That's indeed the reason it fails. I pushed a commit on a separate branch in my private fork that should do the trick: [eboasson@a63a433](https://github.com/eboasson/cyclonedds/commit/a63a433727f659aaf00b0d8bd8db9f988e175da0)

This works!  Thanks.  I have what appears to be sane communication between two terminals running on the guest.  I've run both `ddsperf sanity` between two terminals and a `ddsperf ping` `ddsperf pong` test.  Both seem to work fine.

I have less success between the guest and host, but that's expected, as you've pointed out before.  Similar to the issues reported earlier, the host registers the presence of the guest, but then all subsequent communication times out:

```
# [broomstick] is the host
# [qemu-cheri128-broomstick] is the guest, running on QEMU

bin $ ./ddsperf ping
[7788] participant broomstick:7788: new (self)
[7788] 3.000  rss:4.1MB vcsw:2 ivcsw:0 ddsperf:1%+0%
[7788] participant qemu-cheri128-broomstick:654: new
[7788] 7.000  rss:4.1MB vcsw:7 ivcsw:0 ddsperf:0%+1%
[7788] ping timed out (total 4 times) ... sending new ping
[7788] ping timed out (total 5 times) ... sending new ping
```

I'm not worried about this.  If I can get publishers and subscribers working *within* the guest, that's sufficient for my demo purposes.

**Comment by dodsonmg** [comment]
> Does the same test work fine on Ubuntu?

Yes

> If you can do a packet capture with Wireshark (or you could let Cyclone write a capture file by setting the "Tracing/PacketCaptureFile" option.

I enabled packet capture within a cyclonedds.xml file on both my qemu guest and ubuntu host and ran the combined publisher/subscriber in both places for at least 10 iterations.  I converted the pcap files to text files via `tcpdump -qns 0 -X -r [pcap] > [txt]` and posted them to [this gist](https://gist.github.com/dodsonmg/4f750affd5b30369a505fb6ea46f4b4f).

The qemu guest output (to stdout) looked like this:
```
[INFO] [publisher_node]: Publisher: 'Hello, world! 0'
[INFO] [publisher_node]: Publisher: 'Hello, world! 1'
[INFO] [subscriber_node]: Subscriber: ''
[INFO] [subscriber_node]: Subscriber: ''
[INFO] [publisher_node]: Publisher: 'Hello, world! 2'
[INFO] [subscriber_node]: Subscriber: ''
[INFO] [publisher_node]: Publisher: 'Hello, world! 3'
[INFO] [subscriber_node]: Subscriber: ''
```

The Ubuntu host output (to stdout) looked like this (which matches the expected output):
```
[INFO] [publisher_node]: Publisher: 'Hello, world! 0'
[INFO] [subscriber_node]: Subscriber: 'Hello, world! 0'
[INFO] [publisher_node]: Publisher: 'Hello, world! 1'
[INFO] [subscriber_node]: Subscriber: 'Hello, world! 1'
[INFO] [publisher_node]: Publisher: 'Hello, world! 2'
[INFO] [subscriber_node]: Subscriber: 'Hello, world! 2'
[INFO] [publisher_node]: Publisher: 'Hello, world! 3'
[INFO] [subscriber_node]: Subscriber: 'Hello, world! 3'
```

**Comment by dodsonmg** [comment]
Update:  I modified the publisher/subscriber to just publish and display integers (instead of strings) and it works as expected.

**Comment by eboasson** [comment]
@dodsonmg thanks, that's useful information. It really looks like it is the RMW-specific handling of string ser/deser (and you can exclude the "ser" side if it looks identically starting from the first 4 bytes before the "Hello", which should be on more than the length of the string). I'd happy to look at it if I had some time, but right now time is in unusually limited supply for me ...

@rotu: as you know the ser/deser code we use in the RMW layer well, would it be possible for you to have a look at it? I'm worried it might be something triggered by the presumably-clang-c++-library-instead-of-gnu-c++.

**Comment by rotu** [comment]
I’ll have a look. Not as familiar with the deserialization though.

@dodsonmg what’s the endianness of the publisher and the subscriber machines?

**Comment by dodsonmg** [comment]
The system is FreeBSD-based running on MIPS in emulation on QEMU.  It's big endian.

I'm not actually building `rclpy` because the target doesn't support python (yet).  Anything else I can run instead?

I can run it on the host, but it'll just be Ubuntu running on an x86 machine...

**Comment by rotu** [comment]
Hmm...
Maybe this:
```colcon test --packages-select test_communication --ctest-args \ -L serial```

Also, what are the compile time defines for your build system?
e.g. the output of ```gcc -dM -E -x c /dev/null``` or ```clang -dM -E -x c /dev/null```

I suspect maybe somewhere in the build process, we're using the host system endianness instead of the compilation target endianness.

**Comment by k0ekk0ek** [comment]
> I suspect maybe somewhere in the build process, we're using the host system endianness instead of the compilation target endianness.

That's actually determined in `cyclonedds/src/ddsrt/include/dds/ddsrt/endian.h`, maybe you could try to do a `printf` of `DDSRT_ENDIAN` and see if it matches the platform's endianess(?)

**Comment by rotu** [comment]
@k0ekk0ek I know - that's why I want the preprocesser defines. If DDSRT_ENDIAN is wrong, I want to figure out exactly why :-).

**Comment by k0ekk0ek** [comment]
If @dodsonmg is cross-compiling, I'd expect a CMake toolchain file too. It appears he has a compiler for FreeBSD otherwise `__FreeBSD__` wouldn't have been defined, I'm guessing he's not using a compiler for MIPS? Definitely the right choice to ask for the preprocessor defines. @dodsonmg, maybe include your toolchain file too? Alternatively, instead of cross-compiling, first try compiling on the MIPS target itself (if possible). It'll take a bit longer, but would exclude any cross-compilation issues. If normal compilation works, perhaps then move on to cross-compilation.

**Comment by dodsonmg** [comment]
@k0ekk0ek @rotu 

Here are links to gists for my [CMake toolchain file](https://gist.github.com/dodsonmg/b19442c8aa2a4eebe7e970a4768fe4d7), my [colcon build invocation](https://gist.github.com/dodsonmg/3b46955db0404153fa7d608fc85b4201), and my [preprocessor macros](https://gist.github.com/dodsonmg/511b27c43990dc70bbb5d9df79674b2d).

Compiling within the build environment on QEMU won't work.  I could build a small test file, but trying to build cyclonedds natively on the target on QEMU would a) take hours, and b) run out of memory.

I inserted a `printf` statement in `cyclonedds/src/ddsrt/src/md5.c` and it prints `DDSRT_ENDIAN = 2`. 
 This appears to confirm that it recognises the system is big endian, based on the following code snippet from `cyclonedds/src/ddsrt/include/dds/ddsrt/endian.h` along with my preprocessor macros `#define __BYTE_ORDER__  __ORDER_BIG_ENDIAN__` and `#define __ORDER_BIG_ENDIAN__ 4321`.

```
#define DDSRT_LITTLE_ENDIAN 1
#define DDSRT_BIG_ENDIAN 2

# if defined(__BYTE_ORDER__)
#   if __BYTE_ORDER__ == __ORDER_BIG_ENDIAN__
#     define DDSRT_ENDIAN DDSRT_BIG_ENDIAN
#   elif __BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__
#     define DDSRT_ENDIAN DDSRT_LITTLE_ENDIAN
#   endif
```

**Comment by rotu** [comment]
Perfect! `0000 0000 4865 6c6c` The "Hello, World" is there but the 32-bit length that precedes it is 0 and not `0E` or some mangling of it. This is something to go on!

**Comment by dodsonmg** [comment]
As I said earlier, when I switch from publishing/subscribing `std_msgs::String` to `std_msgs::Int64`, it works fine.

Here is a packet capture from iteration 59.

```
08:14:44.248596 IP 10.0.2.15.29266 > 0.0.0.0.24881: UDP, length 100
        0x0000:  4500 0080 0000 0000 8011 2e5f 0a00 020f  E.........._....
        0x0010:  0000 0000 7252 6131 006c 0000 5254 5053  ....rRa1.l..RTPS
        0x0020:  0201 0110 0110 c451 c61a 43e4 de96 b587  .......Q..C.....
        0x0030:  0900 0008 5e98 13f4 3df4 267b 1504 0020  ....^...=.&{....
        0x0040:  0000 0010 0000 0000 0000 1303 0000 0000  ................
        0x0050:  0000 005a 0000 0000 0000 0000 0000 0059  ...Z...........Y    <--- last byte is 59
        0x0060:  0700 001c 0000 0000 0000 1303 0000 0000  ................
        0x0070:  0000 005a 0000 0000 0000 005a 0000 001f  ...Z.......Z....
```
Similarly, there are only 0-bytes before the `59`, yet the subscriber receives it, processes the message, and displays it correctly.

**Comment by rotu** [comment]
Yes, but the serialization of the string is 32 bit length followed by character data. That length field isn't expected on the Int64 example. Anyway I found the bug. It's my own stupid mistake and I'll fix it post-haste.

**Comment by dodsonmg** [comment]
We've covered a lot of ground here.  Quick summary for future travelers, and then I'll close.

1. Discussion of how to manually enable `ddsperf` (and cyclonedds examples, if desired) without Maven and JDK dependencies.  The patch (for now at least) is available [here](https://github.com/eboasson/cyclonedds/commit/a63a433727f659aaf00b0d8bd8db9f988e175da0).

2. Fixed serialisation of `rmw_cyclonedds` for big-endian systems.  Patch was merged [here](https://github.com/ros2/rmw_cyclonedds/pull/159).  The observed behaviour: a ROS2 publisher was publishing a `std_msgs::String` to a given topic and the subscriber would receive the message but extract an empty string.

3. Discussion of a patch to support socket creation on FreeBSD systems.  Merged in #490.

@k0ekk0ek @rotu Thanks for all the active troubleshooting help and for merging in these changes so quickly.  I really appreciate it.

**Comment by gavanderhoorn** [comment]
I'll just reinforce @dodsonmg's appreciation. @rotu @eboasson @k0ekk0ek super fast turn-around and insightful discussion here. :+1:

**Comment by dodsonmg** [comment]
Whoa, I left @eboasson off my list.  Thanks for taking this question up in the first place and then carrying me to the `rmw` handoff; especially for getting `ddsperf` running on my machine and showing me the wonders of the `cyclonedds.xml` file.

I'm not ungrateful, just dumb.

**Comment by eboasson** [comment]
No worries, @dodsonmg.

I have read up a bit on CHERI and I rather like it, this actually seems truly usable and performant. I know also understand why the `uintptr_t` operations gave you trouble 🙂. I have had a fondness for capability-based security ever since I discovered its existence, and being able to do that at this fine-grained level with limited performance impact while still being able to run "ordinary" C code is very neat.

Hopefully, one day I'll finally do what I have wanted to do for a long time: turn DDS on its head and give it a capability-based security model. That plan got hatched while working on the key ancestor of DDS (called "SPLICE") when I discovered capability-based operating systems. Oh well ... good old times ... we are now two decades later and on the DDS front nothing material has happened ...

**Comment by dodsonmg** [comment]
Funny you should say that.  One of my primary reasons for cross-compiling ROS2 to work on a CHERI system is to explore mapping hardware capabilities to network capabilities.

I'd be glad for a chance to hear more of your thoughts.  Feel free to drop me an email at md403[at]cam.ac.uk.
